streams/src/main/java/org/apache/kafka/streams/state/Stores.java

```diff
@@ -0,0 +1,206 @@
+         * @return the key-value store
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context);
+
+        /**
+         * Create a key value store that stores the given type of key and value, that records changes to a Kafka topic, and that
+         * uses the {@link #withTime(Time) configured Time} instance.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keyClass the class for the keys, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @param valueClass the class for the values, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null, or if the {@code keyClass} or {@code valueClass}
+         *             are not one of {@code String.class},
+         *             {@code Integer.class}, {@code Long.class}, or
+         *             {@code byte[].class}
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Class<K> keyClass, Class<V> valueClass);
+
+        /**
+         * Create a key value store that records changes to a Kafka topic and that uses the {@link #withTime(Time) configured
+         * Time} instance and the supplied serializers and deserializers.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keySerializer the serializer for keys; may not be null
+         * @param keyDeserializer the deserializer for keys; may not be null
+         * @param valueSerializer the serializer for values; may not be null
+         * @param valueDeserializer the deserializer for values; may not be null
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Serializer<K> keySerializer, Deserializer<K> keyDeserializer,
+                                                 Serializer<V> valueSerializer, Deserializer<V> valueDeserializer);
+    }
+
+    /**
+     * The factory for creating in-memory key-value stores.
+     */
+    public static interface InMemoryFactory extends KeyValueStoreFactory {
```

Comment:
What I meant is that InMemoryFactory / OffHeapFactory should probably not extending from KeyValueStoreFactory since from their names they are more general factories; for example, in the future you may have a DocumentStore / etc, and an InMemoryDocumentStore implementing this interface. InMemoryFactory will provide general functionalities like capacity limits you mentioned, but not limited to key-value store.

Now since we only have key-value stores, Stores.inMemory will use in-memory key-value stores "by default"; but moving forward we may want Stores.inMemory().keyValue().create() to separate different types. And if we follow this path then I think InMemoryFactory / OffHeapFactory should not extend from a specific KeyValueStoreFactory.